applications/tari_dan_app_utilities/src/template_manager/implementation/manager.rs

```diff
@@ -301,10 +350,34 @@ impl<TAddr: NodeAddressable + Send + Sync + 'static> TemplateProvider for Templa
             return Ok(Some(template));
         }
 
-        let Some(template) = self.fetch_template(address).optional()? else {
+        let Some(template_result) = self.fetch_template(address).optional()? else {
             return Ok(None);
         };
+
         debug!(target: LOG_TARGET, "CACHE MISS: Template {}", address);
+
+        // getting template
+        let template = match template_result {
+            TemplateResult::Template(template) => template,
+            TemplateResult::PendingTemplate => {
+                let start = SystemTime::now();
+                let mut template = None;
+                while template.is_none() {
+                    let elapsed = start.duration_since(UNIX_EPOCH).expect("Time went backwards");
+                    if elapsed.gt(&self.config.pending_templates_wait_timeout()) {
+                        break;
+                    }
+                    if let Some(TemplateResult::Template(fetched_template)) = self.fetch_template(address).optional()? {
+                        template = Some(fetched_template);
+                    }
+                    // sleeping here to not overload the local database while waiting for the template to be ready
+                    thread::sleep(std::time::Duration::from_millis(100));
```

Comment:
Not ideal. Will think about how we should handle pending templates in a transaction.